playground/backend/internal/cloud_bucket/precompiled_objects.go

```diff
@@ -0,0 +1,283 @@
+	client, err := storage.NewClient(ctx, option.WithoutAuthentication())
+	if err != nil {
+		return nil, fmt.Errorf("storage.NewClient: %v", err)
+	}
+	defer client.Close()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*Timeout)
+	defer cancel()
+
+	precompiledObjects := make(SdkToCategories, 0)
+	bucket := client.Bucket(BucketName)
+
+	dirs, err := cd.getPrecompiledObjectsDirs(ctx, targetSdk, bucket)
+	if err != nil {
+		return nil, err
+	}
+
+	for objectDir := range dirs {
+		infoPath := filepath.Join(objectDir, MetaInfoName) // helping file with information about this object
+		rc, err := bucket.Object(infoPath).NewReader(ctx)
+		if err != nil {
+			logger.Errorf("Object(%q).NewReader: %v", infoPath, err.Error())
+			continue
+		}
+		data, err := ioutil.ReadAll(rc)
+		if err != nil {
+			logger.Errorf("ioutil.ReadAll: %v", err.Error())
+			continue
+		}
+		precompiledObject := ObjectInfo{}
+		err = json.Unmarshal(data, &precompiledObject)
+		if err != nil {
+			logger.Errorf("json.Unmarshal: %v", err.Error())
+			continue
+		}
+		for _, objectCategory := range precompiledObject.Categories {
+			if targetCategory == "" || targetCategory == objectCategory { //take only requested categories
+				appendPrecompiledObject(precompiledObject, &precompiledObjects, objectDir, objectCategory)
+			}
+		}
+		rc.Close()
+	}
+	return &precompiledObjects, nil
+}
+
+// GetCategoryToPrecompiledObjects adds categories with precompiled objects to protobuf object
+func GetCategoryToPrecompiledObjects(categoryName string, precompiledObjects *PrecompiledObjects, sdkCategories *pb.Categories) {
+	category := pb.Categories_Category{
+		CategoryName:       categoryName,
+		PrecompiledObjects: make([]*pb.PrecompiledObject, 0),
+	}
+	for _, object := range *precompiledObjects {
+		category.PrecompiledObjects = append(category.PrecompiledObjects, &pb.PrecompiledObject{
+			CloudPath:   object.CloudPath,
+			Name:        object.Name,
+			Description: object.Description,
+			Type:        object.Type,
+		})
+	}
+	sdkCategories.Categories = append(sdkCategories.Categories, &category)
+}
+
+// getPrecompiledObjectsDirs finds "directories" with precompiled objects
+func (cd *CloudStorage) getPrecompiledObjectsDirs(ctx context.Context, targetSdk pb.Sdk, bucket *storage.BucketHandle) (map[string]bool, error) {
+	prefix := targetSdk.String()
+	if targetSdk == pb.Sdk_SDK_UNSPECIFIED {
+		prefix = ""
+	}
+	it := bucket.Objects(ctx, &storage.Query{
+		Prefix: prefix,
+	})
+	objectDirs := make(map[string]bool, 0)
+	for {
+		attrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("Bucket(%q).Objects: %v", BucketName, err)
+		}
+		path := attrs.Name
+		if isPathToPrecompiledObject(path) {
+			precompiledObjectDir := filepath.Dir(path)
+			_, ok := objectDirs[precompiledObjectDir]
+			if !ok {
+				objectDirs[precompiledObjectDir] = true
```

Comment:
Could you please add a comment for this logic? I don't quite get why we have such a map with boolean values.

Reply:
because it is a common way instead of a set. There are no sets in Goland

Reply:
that's fine

Reply:
I don't talk about sets, I don't get the whole logic at all, it's very implicit – that's why I'm asking to add comments for it to describe what's going on